tests/system/src/test/java/test/util/ScreenCaptureTestWatcher.java

```diff
@@ -0,0 +1,114 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package test.util;
+
+import java.awt.image.BufferedImage;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.imageio.ImageIO;
+import javafx.embed.swing.SwingFXUtils;
+import javafx.geometry.Rectangle2D;
+import javafx.scene.image.WritableImage;
+import javafx.scene.robot.Robot;
+import javafx.stage.Screen;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestWatcher;
+
+/**
+ * Standard Test Watcher for Headful Tests (JUnit 5 Only).
+ * <p>
+ * This facility takes a screenshot of any failed test, then logs the base64-encoded screenshot
+ * to {@code stderr}.
+ * <p>
+ * To use, simply add the following annotation to your class:
+ * <pre>{@code
+ * @ExtendWith(ScreenCaptureTestWatcher.class)
+ * }</pre>
+ * <p>
+ * For Eclipse users, override dependencies with the following entries:
+ * <pre>{@code
+ * --add-modules=javafx.base,javafx.graphics,javafx.controls,javafx.swing
+ * --add-opens javafx.controls/test.com.sun.javafx.scene.control.test=javafx.base
+ * --add-exports javafx.base/com.sun.javafx=ALL-UNNAMED
+ * -Djava.library.path="../../../../build/sdk/lib"
+ * -ea
+ * }</pre>
+ * <p>
+ * WARNING: using this utility may significantly increase the size of your logs!
+ * Make sure there is plenty of free disk space.
+ */
+// TODO investigate having a hard-coded or programmable via property limit on the number of
+// captured screenshots.
+public class ScreenCaptureTestWatcher implements TestWatcher {
+    @Override
+    public void testAborted(ExtensionContext extensionContext, Throwable err) {
+        err.printStackTrace();
+    }
+
+    @Override
+    public void testFailed(ExtensionContext extensionContext, Throwable err) {
+        err.printStackTrace();
+        System.err.println(generateScreenshot("Screenshot:{", "}"));
```

Comment:
When does the `testFailed` method run? After a failing `@Test` method that throws the exception? What if there is more than one failing test? Ideally what we want is something that runs after all failing tests, but before the `@AfterEach` method.

Reply:
Or... maybe we really do want to take a screen dump for each failing test in a test class.

Also, what happens if one of the lifecycle methods is the one that throws the error (e.g., `Before` `After`, etc)?

Reply:
good question!  A quick test indicates that `@BeforeAll` and `@AfterAll` are treated as the framework errors and not handled by the watcher, but the failures in `@Test`, `@BeforeEach`, and `@AfterEach` are sent to the test watcher.